src/openassetio-core/include/openassetio/specification/Specification.hpp

```diff
@@ -0,0 +1,42 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2013-2022 The Foundry Visionmongers Ltd
+/**
+ * Provide the base dynamic specification class.
+ */
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "trait/property.hpp"
+
+namespace openassetio {
+inline namespace OPENASSETIO_VERSION {
+namespace specification {
+
+/**
+ * Base specification class that is defined by a given list of trait
+ * IDs.
+ */
+class OPENASSETIO_CORE_EXPORT Specification {
+ public:
+  using TraitIds = std::vector<trait::TraitId>;
+
+  explicit Specification(const TraitIds& traitIds);
+  virtual ~Specification();
+
+  [[nodiscard]] bool hasTrait(const trait::TraitId& traitId) const;
+
+  [[nodiscard]] bool getTraitProperty(trait::property::Value* out, const trait::TraitId& traitId,
```

Comment:
[question] Do we want to consider typed accessors here?

Reply:
We can move the typed accessor from `TraitBase` to here.  I was in two minds where it belongs. 

I guess I viewed `getTraitProperty` as an advanced method that normal users would never call, since they should be using trait views, so didn't want to be too friendly.

The wrapper typed accessor in `TraitBase` is also `protected`, reducing the public API surface area, whereas if we move it here it'll have to be `public`.

Do you think it's worth adding as part of the public interface?

Reply:
Capturing off-line discussion:

For now, the main choice is: do we leak `std::variant` in the public API?

Reply:
We can certainly save this for a later PR if we're happy that we're leaking `std::variant` in the public API.

Reply:
Do we need to chat about this some more?  It sounds like we do want to expose `std::variant`, along with (eventually) a `std::optional` API, to make the modern C++ folks happy?

But if you think we should hide it for now we can do that.  Hiding `std::variant` makes the Python bindings a bit trickier - pybind is clever enough to work with `std::variant` out of the box, so if we hide it's use behind a typed accessor then we'll need a way to "unhide" it (or equivalent) in the bindings.

Reply:
One argument perhaps is that it's just coincidental that we return a `std::variant` as well as use them internally.  We can have an API that returns `std::variant` and swap out the underlying storage for something else, they're not necessarily coupled.

Reply:
Shall we leave it as a variant for now, and pick this up in another ticket in the not too distant?